In [1]:
import pandas as pd

# import sys
# sys.path.append("../../src/wikipedia")

# from wikipedia_parser import process_wikipedia_dump


# dump_file = "../../data/wikipedia/is/iswiki-20240901-pages-articles.xml"
# df = process_wikipedia_dump(dump_file, max_pages=None)
# df.to_parquet("iswiki.processed_datadump.parquet")

df = pd.read_parquet("iswiki.processed_datadump.parquet")

print(len(df))
df.head()


86163


page_id      title                                                url  \
0       3   Finnland             https://is.wikipedia.org/wiki/Finnland   
1       4    Forsíða    https://is.wikipedia.org/wiki/Fors%C3%AD%C3%B0a   
2       6  Reykjavík       https://is.wikipedia.org/wiki/Reykjav%C3%ADk   
3       7    Svíþjóð  https://is.wikipedia.org/wiki/Sv%C3%AD%C3%BEj%...   
4     591    Danmörk         https://is.wikipedia.org/wiki/Danm%C3%B6rk   

   word_count  outlink_count  category_count categories  template_count  \
0        4306            553               0                        110   
1          28              5               0                         11   
2        2323            309               0                         51   
3        3025            360               0                         44   
4        4711            489               0                         55   

   external_link_count         last_modified  \
0                   47  2024-04-18T10:18:03Z   
1                    0  2021-01-02T00:52:18Z   
2                   41  2024-04-09T12:51:22Z   
3                   56  2024-08-12T16:35:20Z   
4                   58  2024-07-18T11:56:58Z   

                                      processed_text  \
0  \nFinnland (finnska Suomi; sænska: Finland), f...   
1  \n\n\n\n\n\n\n\nGrein mánaðarins\n\nEldri grei...   
2  \n\nReykjavík er höfuðborg Íslands, fjölmennas...   
3  \n\nSvíþjóð (sænska: Sverige), formlegt heiti ...   
4  \nDanmörk (danska: Danmark; ) er land í Norður...   

                                            raw_text  
0  {{Land\n| nafn = Lýðveldið Finnland\n| nafn_á_...  
1  <templatestyles src="Template:Forsíða/styles.c...  
2  {{Um|höfuðborg Íslands|aðrar merkingar orðsins...  
3  {{Land \n|nafn = Konungsríkið Svíþjóð\n|nafn_á...  
4  {{Land\n| nafn = Konungsríkið Danmörk\n| nafn_...

In [2]:
# df_pageviews = pd.read_parquet("202309_monthly_views.parsed.parquet")

pageviews_file = "/Users/einar/git/hafsteinn/together_rag/data/wikipedia_meta/aggregated/is.wikipedia_monthly_views.parquet"
df_pageviews = pd.read_parquet(pageviews_file)
# print(len(df_pageviews))

print(f"Number of pageviews rows: {len(df_pageviews):,}")

df_pageviews.head(2)


Number of pageviews rows: 1,182,149


wiki_code            article_title  views page_ids
0  is.wikipedia  "\"Weird_Al\"_Yankovic"      1  [52652]
1  is.wikipedia                        $      2  [30714]

**have almos 1.2 million rows in the pageviews data.**

~~This is interesting since according to the frontpage of is.wikipedia.com we have almost **60K** articles:~~  
This is because we first need to group by the article title, have many entries of the same title.

```
Á hinni íslensku Wikipedíu eru nú 59.115 greinar.
```




## Lets try to transform to page_ids 

I.e. now we have article_title which can have a list of page_ids.

I want to see how it would look if we have page_id to a list of article titles instead..



In [5]:
df_pageviews.head(2)

wiki_code            article_title  views page_ids
0  is.wikipedia  "\"Weird_Al\"_Yankovic"      1  [52652]
1  is.wikipedia                        $      2  [30714]

In [6]:
from collections import defaultdict

d = defaultdict(list)

for index, row in df_pageviews.iterrows():
    for page_id in row.page_ids:
        # d[page_id].append(row.article_title)
        # also need the wiki_code and the views column
        d[page_id].append((row.wiki_code, row.article_title, row.views))


print(len(d))

105702


In [ ]:
# df_pageviews_by_page_id = pd.DataFrame(list(d.items()), columns=["page_id", "article_titles"]
df_pageviews_by_page_id = pd.DataFrame(d.values(), columns=["wiki_code", "article_title", "views"])
df_pageviews_by_page_id.head(2)


### Explore the data

In [7]:
df_pageviews = df_pageviews[df_pageviews.wiki_code == "is.wikipedia"]

print(len(df_pageviews))
df_pageviews.head(2)

1182149


wiki_code            article_title  views page_ids
0  is.wikipedia  "\"Weird_Al\"_Yankovic"      1  [52652]
1  is.wikipedia                        $      2  [30714]

### How many of the pages can we match by title

In [8]:
len(df_pageviews) - len(df)

1095986

There are more rows in the pageviews data, we should hopefully 
have data for each of the entries in the processed dump.



In [9]:
xpageviews_titles = set(df_pageviews.article_title)

print(f"Number of titles in data: {len(data_titles)}")
print(f"Number of titles in pageviews: {len(pageviews_titles)}")
print(f"Number of titles in both data: {len(data_titles.intersection(pageviews_titles))}")


Number of titles in data: 86163
Number of titles in pageviews: 202415
Number of titles in both data: 40861


In [10]:
df[~df.title.isin(df_pageviews.article_title)].head(2)

page_id                     title  \
7     599           Föll í íslensku   
9     602  Borgarstjóri Reykjavíkur   

                                                 url  word_count  \
7  https://is.wikipedia.org/wiki/F%C3%B6ll_%C3%AD...         148   
9  https://is.wikipedia.org/wiki/Borgarstj%C3%B3r...         683   

   outlink_count  category_count categories  template_count  \
7             11               0                          1   
9            191               0                          4   

   external_link_count         last_modified  \
7                    0  2023-09-12T13:13:18Z   
9                    5  2024-08-16T13:27:12Z   

                                      processed_text  \
7  \n\nÍ íslensku eru fjögur föll, nefnifall, þol...   
9  \nBorgarstjóri Reykjavíkur er æðsti embættisma...   

                                            raw_text  
7  {{Föll í íslensku}}\n\nÍ [[Íslenska|íslensku]]...  
9  [[Mynd:20190624 CityHall 9152 (48463162702).jp...

So page_id 602 with title "Borgarstjóri Reykjavíkur" is not in the pageviews data.


In [7]:
df_pageviews[df_pageviews.article_title.str.contains("Borgarstjóri", case=False)]

wiki_code                  article_title  views     page_ids
28456220  is.wikipedia                   Borgarstjóri     17      [51093]
28456221  is.wikipedia       Borgarstjóri_Reykjavíkur     78  [602, null]
28456222  is.wikipedia                 Borgarstjórinn      4     [159481]
28517568  is.wikipedia  Páll_Einarsson_(borgarstjóri)     16      [65350]

It is a simple case of pageviews data having _ instead of space. Also we might be able to use the page id

In [11]:
df_pageviews["title_match"] = df_pageviews.article_title.str.replace("_", " ").str.lower()
df["title_match"] = df.title.str.replace("_", " ").str.lower()

data_titles = set(df.title_match)
pageviews_titles = set(df_pageviews.title_match)

print(f"Number of titles in data: {len(data_titles)}")
print(f"Number of titles in pageviews: {len(pageviews_titles)}")
print(f"Number of titles in both data: {len(data_titles.intersection(pageviews_titles))}")
print(f"Number of unmatched titles in data: {len(data_titles) - len(data_titles.intersection(pageviews_titles))}")
print(f"Number of unmatched titles in pageviews: {len(pageviews_titles) - len(data_titles.intersection(pageviews_titles))}")


Number of titles in data: 85506
Number of titles in pageviews: 200264
Number of titles in both data: 85148
Number of unmatched titles in data: 358
Number of unmatched titles in pageviews: 115116


### still many unmatched titles

But way less than before, there are still many unmatched titles.

In [9]:
pd.set_option("display.max_colwidth", 100)
df[~df.title_match.isin(df_pageviews.title_match)].sort_values(by=["external_link_count"], ascending=False).head(5)

page_id                              title  \
83325  177298         Forsetakjör á Íslandi 2024   
61256  141279  Borgarastyrjöldin í Jemen (2014–)   
85534  181102                   Peter Pellegrini   
84247  178822                       Alsírstríðið   
45794  106161         Forsetakjör á Íslandi 2012   

                                                                                         url  \
83325                https://is.wikipedia.org/wiki/Forsetakj%C3%B6r_%C3%A1_%C3%8Dslandi_2024   
61256  https://is.wikipedia.org/wiki/Borgarastyrj%C3%B6ldin_%C3%AD_Jemen_%282014%E2%80%93%29   
85534                                         https://is.wikipedia.org/wiki/Peter_Pellegrini   
84247                         https://is.wikipedia.org/wiki/Als%C3%ADrstr%C3%AD%C3%B0i%C3%B0   
45794                https://is.wikipedia.org/wiki/Forsetakj%C3%B6r_%C3%A1_%C3%8Dslandi_2012   

       word_count  outlink_count  category_count categories  template_count  \
83325        3485            306               0                         64   
61256        1827            172               0                        130   
85534         724             72               0                         40   
84247        2148            178               0                        126   
45794         761             44               0                         19   

       external_link_count         last_modified  \
83325                   90  2024-07-06T11:59:01Z   
61256                   75  2024-08-30T23:20:10Z   
85534                   61  2024-06-19T00:15:13Z   
84247                   59  2024-05-03T02:04:25Z   
45794                   52  2024-07-06T11:08:17Z   

                                                                                            processed_text  \
83325  \nForsetakjör á Íslandi 2024 fór fram þann 1. júní 2024. Halla Tómasdóttir sigraði með 34,1% gil...   
61256  \nBorgarastyrjöldin í Jemen eru átök sem hófust árið 2014 og eru tveir hópar sem berjast þar um ...   
85534  \nPeter Pellegrini (f. 6. október 1975) er slóvakískur stjórnmálamaður og núverandi forseti Slóv...   
84247  \nAlsírstríðið, einnig kallað alsírska sjálfstæðisstríðið eða alsírska byltingin (arabíska: الثو...   
45794  thumb|right|Híbýli forseta Íslands eru á Bessastöðum.\nForsetakjör á Íslandi 2012 fór fram lauga...   

                                                                                                  raw_text  \
83325  [[Mynd:Forsetakosningar 2024.svg|thumb|Niðurstöður kosninganna]]\n'''Forsetakjör á Íslandi 2024'...   
61256  {{Infobox military conflict\n| conflict    = Borgarastyrjöldin í Jemen\n| partof      = [[Arabís...   
85534  {{Forseti\n| nafn = Peter Pellegrini\n| mynd           = Peter Pellegrini, 2024 (cropped).jpg\n|...   
84247  {{Infobox military conflict\n| conflict          = Alsírstríðið<br />{{Nobold|{{lang|ar|ثورة الت...   
45794  [[Mynd:OB090126c-3053_Bessastadir.JPG|thumb|right|Híbýli forseta Íslands eru á [[Bessastaðir|Bes...   

                             title_match  
83325         forsetakjör á íslandi 2024  
61256  borgarastyrjöldin í jemen (2014–)  
85534                   peter pellegrini  
84247                       alsírstríðið  
45794         forsetakjör á íslandi 2012

In [10]:
df.page_id.dtype

dtype('O')

In [11]:
unmatched_ids = df[~df.title_match.isin(df_pageviews.title_match)].page_id.unique()
len(unmatched_ids)

23919

In [12]:
type(df_pageviews.iloc[0].page_ids[0])

str

In [13]:
# df_pageviews[df_pageviews.page_id.isin(unmatched_ids)].head(2)
# the page_ids column in df_pageviews is a list of page ids, so we need to match them to the page_id in df

# def check_page_id_match(page_ids, unmatched_ids):
#     return any(str(id) in page_ids for id in unmatched_ids)

# matched_rows = df_pageviews[df_pageviews['page_ids'].apply(lambda x: check_page_id_match(x, unmatched_ids))]

# print(f"Number of matched rows: {len(matched_rows)}")
# matched_rows.head()

import numpy as np

# Convert unmatched_ids to a set of strings for faster lookup
unmatched_ids_set = set(map(str, unmatched_ids))

# Vectorized function to check for matches
def check_match(page_ids):
    return np.frompyfunc(lambda x: x in unmatched_ids_set, 1, 1)(page_ids).any()

# Apply the vectorized function
matched_mask = df_pageviews['page_ids'].apply(check_match)

# Get the matched rows
matched_rows = df_pageviews[matched_mask]

print(f"Number of matched rows: {len(matched_rows)}")
matched_rows.head()




Number of matched rows: 506


wiki_code article_title  views  \
28446992  is.wikipedia             -    352   
28449654  is.wikipedia     621_f.Kr.      1   
28449656  is.wikipedia     622_f.Kr.      2   
28449657  is.wikipedia     623_f.Kr.      1   
28449659  is.wikipedia     626_f.Kr.      1   

                                                                                                     page_ids  \
28446992  [65519, 155031, 171906, 33641, 127265, 20264, 171929, 2453, 27685, 18332, 26113, 25992, 132275, ...   
28449654                                                                                             [145150]   
28449656                                                                                             [145150]   
28449657                                                                                             [145150]   
28449659                                                                                             [145150]   

         title_match  
28446992           -  
28449654   621 f.kr.  
28449656   622 f.kr.  
28449657   623 f.kr.  
28449659   626 f.kr.

In [14]:
debug_ids = [65519, 155031, 171906, 33641, 127265, 20264, 171929, 2453, 27685, 18332, 26113, 25992, 132275
]
debug_ids = [str(id) for id in debug_ids]

df[df.page_id.isin(debug_ids)]







page_id                               title  \
881      2453                            Hannover   
6560    18332                           Christina   
8268    20264                           Pokabjörn   
10999   25992                          Stonehenge   
11065   26113                         Flokkskvísl   
11586   27685  Bandamenn (seinni heimsstyrjöldin)   
14391   33641                         Frida Kahlo   
28479   65519                           Matarhefð   
54036  127265          Stríð Rússlands og Úkraínu   
55644  132275                         Zootropolis   
69326  155031                    Alþjóðasamskipti   

                                                                                     url  \
881                                               https://is.wikipedia.org/wiki/Hannover   
6560                                             https://is.wikipedia.org/wiki/Christina   
8268                                        https://is.wikipedia.org/wiki/Pokabj%C3%B6rn   
10999                                           https://is.wikipedia.org/wiki/Stonehenge   
11065                                     https://is.wikipedia.org/wiki/Flokkskv%C3%ADsl   
11586          https://is.wikipedia.org/wiki/Bandamenn_%28seinni_heimsstyrj%C3%B6ldin%29   
14391                                          https://is.wikipedia.org/wiki/Frida_Kahlo   
28479                                       https://is.wikipedia.org/wiki/Matarhef%C3%B0   
54036  https://is.wikipedia.org/wiki/Str%C3%AD%C3%B0_R%C3%BAsslands_og_%C3%9Akra%C3%ADnu   
55644                                          https://is.wikipedia.org/wiki/Zootropolis   
69326                      https://is.wikipedia.org/wiki/Al%C3%BEj%C3%B3%C3%B0asamskipti   

       word_count  outlink_count  category_count categories  template_count  \
881          2088            162               0                         12   
6560          785              2               0                          4   
8268          330             38               0                          3   
10999         168             21               0                          1   
11065          50              4               0                          0   
11586         500            103               0                          1   
14391         411             31               0                          3   
28479          57             21               0                          1   
54036         291             38               0                          9   
55644         214             61               0                          4   
69326          65             14               0                          1   

       external_link_count         last_modified  \
881                      0  2022-11-28T17:19:54Z   
6560                     2  2023-08-07T00:11:27Z   
8268                     2  2022-03-26T04:50:17Z   
10999                    0  2024-01-31T10:46:26Z   
11065                    0  2013-03-08T14:34:58Z   
11586                    0  2024-04-10T11:59:13Z   
14391                    0  2024-07-09T01:08:56Z   
28479                    0  2013-07-03T09:28:26Z   
54036                    6  2024-08-17T10:41:54Z   
55644                    1  2023-09-24T10:49:47Z   
69326                    0  2019-11-26T18:18:47Z   

                                                                                            processed_text  \
881    \nHannover er höfuðborg og stærsta borg þýska sambandslandsins Neðra-Saxlands með um 540.000 íbú...   
6560   \nChristina er íslenskt kvenmannsnafn.\n\n== Dreifing á Íslandi ==\n\n\nImageSize = width:600 he...   
8268   \nPokabjörn (eða kóalabjörn) (fræðiheiti: Phascolarctos cinereus) er spendýr af pokadýraflokki s...   
10999  \n\nStonehenge er mannvirki frá nýsteinöld og bronsöld, gert úr risasteinum staðsett við Amesbur...   
11065  Flokkskvísl, sem gengur einnig undir nöfnunum riðill, sveit og lið, er hernaðareining sem samans...   
11586  \nBandamenn í Síðari he

In [15]:
s1 = "(What's The Story) Morning Glory?"
s2 = "(What's the Story) Morning Glory?"